service/history/shard/context.go

```diff
@@ -1170,6 +1172,35 @@ func (s *contextImpl) GetLastUpdatedTime() time.Time {
 	return s.lastUpdated
 }
 
+func (s *contextImpl) InsertFailoverMarkers(
+	markers []*persistence.FailoverMarkerTask,
+) error {
+
+	tasks := make([]persistence.Task, 0, len(markers))
+	for _, marker := range markers {
+		tasks = append(tasks, marker)
+	}
+
+	s.Lock()
+	defer s.Unlock()
+
+	transferMaxReadLevel := int64(0)
+	if err := s.allocateTransferIDsLocked(
+		tasks,
+		&transferMaxReadLevel,
+	); err != nil {
+		return err
+	}
+	defer s.updateMaxReadLevelLocked(transferMaxReadLevel)
+
+	if err := s.executionManager.CreateFailoverMarkerTasks(&persistence.CreateFailoverMarkersRequest{
```

Comment:
1. Maybe add some retry logic here? 
2. I am not sure about this but do we need to ensure the shard ownership is not lost when inserting the marker tasks?